presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java

```diff
@@ -210,27 +220,43 @@ public boolean isFinished()
     public Page getNextPage()
     {
         try {
-            Block[] blocks = new Block[columnMappings.size()];
             Page dataPage = delegate.getNextPage();
             if (dataPage == null) {
                 return null;
             }
+
+            if (bucketAdapter.isPresent()) {
+                IntArrayList ids = bucketAdapter.get().computeEligibleRowIds(dataPage);
+                Block[] adaptedBlocks = new Block[dataPage.getChannelCount()];
+                for (int i = 0; i < adaptedBlocks.length; i++) {
+                    adaptedBlocks[i] = new LazyBlock(ids.size(), new RowFilterLazyBlockLoader(dataPage.getBlock(i), ids));
```

Comment:
if the block is lazy and already loaded, we don't need to wrapper with a lazy block, and instead we can just mask here

Reply:
I believe the code in `computeEligibleRowIds(dataPage)` forces some of the channels to load.  For those channels, you don't need to create a new `LazyBlock` instead something like:
````
Block block = dataPage.getBlock(i);
if (block instanceof LazyBlock && ((LazyBlock) block).isLoaded()) {
    adaptedBlocks[i] = maskBlock(ids, ((LazyBlock) block).getBlock());
} 
else {
    adaptedBlocks[i] = new LazyBlock(ids.size(), new RowFilterLazyBlockLoader(dataPage.getBlock(i), ids));
}

Reply:
The code you suggested appears to be wrong. But thanks for the code, it's very helpful in helping me understand things.